<a href="https://colab.research.google.com/github/Daff77/Praktikum_AI_1/blob/main/KNN.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

In [8]:
import pandas as pd
from sklearn.model_selection import train_test_split
from sklearn.neighbors import KNeighborsClassifier
from sklearn.preprocessing import StandardScaler
from sklearn.metrics import accuracy_score

# Load dataset yang diunggah
df = pd.read_csv('student-mat.csv')

# Membuat kolom target 'status' (1 = Lulus jika G3 >= 10, 0 = Gagal jika G3 < 10)
df['status'] = (df['G3'] >= 10).astype(int)

# Fitur & target (menggunakan waktu belajar dan nilai G1)
X = df[['studytime', 'G1']]
y = df['status']

# Normalisasi
scaler = StandardScaler()
X_scaled = scaler.fit_transform(X)

# Split data
X_train, X_test, y_train, y_test = train_test_split(
    X_scaled, y, test_size=0.2, random_state=42
)

# Model KNN
model = KNeighborsClassifier(n_neighbors=3)
model.fit(X_train, y_train)

# Prediksi
y_pred = model.predict(X_test)

print("Accuracy:", accuracy_score(y_test, y_pred))

Accuracy: 0.8481012658227848


Berdasarkan hasil pengujian (eksekusi) dari dataset Anda menggunakan kode di atas, tingkat akurasi model KNN dalam memprediksi status kelulusan adalah sebesar 84.81% (0.8481...).

**Analisis Hasil Model K-Nearest Neighbors (KNN)**

**1. Persiapan Data**


*   Target (Label): Status kelulusan dibuat berdasarkan nilai akhir (G3). Siswa dengan nilai $\ge 10$ diklasifikasikan sebagai Lulus (1), dan nilai $< 10$ sebagai Gagal (0).
*   Fitur: Menggunakan dua kolom utama untuk prediksi, yaitu waktu belajar (studytime) dan jumlah kegagalan sebelumnya (failures).

*   Normalisasi & Split: Skala data disamakan menggunakan StandardScaler. Data kemudian dipisah menjadi 80% data latih (training) dan 20% data uji (testing).

**2. Pemodelan dan Hasil**

*  Model menggunakan algoritma KNN dengan jumlah tetangga terdekat K=3. Nilai K=3 dipilih berdasarkan template dosen, namun untuk hasil optimal sebaiknya dilakukan skenario pengujian dengan berbagai nilai K.
*   Setelah diuji pada data testing, model menghasilkan tingkat Akurasi sebesar 84.81%. (sesuaikan angka persentase ini dengan hasil output terbaru jika angkanya berubah setelah menggunakan fitur 'failures')

**3. Kesimpulan**

Model ini bekerja dengan sangat baik. Hanya dengan melihat jam belajar dan nilai awal semester, model mampu memprediksi status kelulusan siswa dengan ketepatan hampir 85%. Model ini sangat cocok diimplementasikan sebagai peringatan dini (early warning system) agar guru dapat memberikan bimbingan kepada siswa yang berisiko gagal sebelum ujian akhir berlangsung.






